src/Avalonia.Visuals/Utility/ReadOnlySlice.cs

```diff
@@ -0,0 +1,167 @@
+
+        /// <summary>
+        ///     Returns a span of the underlying buffer.
+        /// </summary>
+        /// <returns>The <see cref="ReadOnlySpan{T}"/> of the underlying buffer.</returns>
+        public ReadOnlySpan<T> AsSpan()
+        {
+            return Buffer.Span.Slice(Start, Length);
+        }
+
+        /// <summary>
+        ///     Returns a sub slice of elements that start at the specified index and has the specified number of elements.
+        /// </summary>
+        /// <param name="start">The start of the sub slice.</param>
+        /// <param name="length">The length of the sub slice.</param>
+        /// <returns>A <see cref="ReadOnlySlice{T}"/> that contains the specified number of elements from the specified start.</returns>
+        public ReadOnlySlice<T> AsSlice(int start, int length)
+        {
+            if (start < 0 || start >= Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(start));
+            }
+
+            if (Start + start > End)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            return new ReadOnlySlice<T>(Buffer, Start + start, length);
+        }
+
+        /// <summary>
+        ///     Returns a specified number of contiguous elements from the start of the slice.
+        /// </summary>
+        /// <param name="length">The number of elements to return.</param>
+        /// <returns>A <see cref="ReadOnlySlice{T}"/> that contains the specified number of elements from the start of this slice.</returns>
+        public ReadOnlySlice<T> Take(int length)
+        {
+            if (length > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            return new ReadOnlySlice<T>(Buffer, Start, length);
+        }
+
+        /// <summary>
+        ///     Bypasses a specified number of elements in the slice and then returns the remaining elements.
+        /// </summary>
+        /// <param name="length">The number of elements to skip before returning the remaining elements.</param>
+        /// <returns>A <see cref="ReadOnlySlice{T}"/> that contains the elements that occur after the specified index in this slice.</returns>
+        public ReadOnlySlice<T> Skip(int length)
+        {
+            if (length > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            return new ReadOnlySlice<T>(Buffer, Start + length, Length - length);
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            return new ReadOnlySliceEnumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        int IReadOnlyCollection<T>.Count => Length;
+
+        T IReadOnlyList<T>.this[int index] => this[index];
+
+        private class ReadOnlySliceEnumerator : IEnumerator<T>
```

Comment:
Since this is supposed to be general purpose I would suggest implementing a struct enumerator instead.

Reply:
Y will do so